### ISSUE TITLE
Re-resolve hostnames as fallback when all hosts are unreachable

### ISSUE THREAD

**Issue by wprzytula** [scylladb/java-driver#224]
If all nodes in the cluster change their IPs at one time, driver can no longer ever contact the cluster; the only solution is to restart the driver. A fallback should be added to the control connection reconnect logic so that when no known host is reachable, all hostnames among initial contact points are reresolved and control connection is attempted to be opened to any of them. If this succeeds, a metadata fetch is issued normally and the whole cluster is discovered with its new IPs.

An analogous feature has been merged to gocql lastly: https://github.com/gocql/gocql/pull/1708

Fix: https://github.com/scylladb/java-driver/issues/224

Migrated to Jira: [DRIVER-199](https://scylladb.atlassian.net/browse/DRIVER-199)

[DRIVER-199]: https://scylladb.atlassian.net/browse/DRIVER-199?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by roydahan** [comment]
In scylla-4.x we should have the parameter called advanced.resolve-contact-points which should take care of that.
So, if it works this enhancement should be done only for scylla-3.x

**Comment by mykaul** [comment]
This needs to get some attention. It was solved in most drivers already, not in Java 3.x.

**Comment by dkropachev** [comment]
It is blocked by https://github.com/scylladb/java-driver/issues/333 and https://github.com/scylladb/java-driver/issues/334

**Comment by mykaul** [comment]
> It is blocked by [#333](https://github.com/scylladb/java-driver/issues/333) and [#334](https://github.com/scylladb/java-driver/issues/334)

#333 is now fixed, what about #334  ?

**Comment by dkropachev** [comment]
> > It is blocked by [#333](https://github.com/scylladb/java-driver/issues/333) and [#334](https://github.com/scylladb/java-driver/issues/334)
> 
> [#333](https://github.com/scylladb/java-driver/issues/333) is now fixed, what about [#334](https://github.com/scylladb/java-driver/issues/334) ?

Both done, this issue is ublocked.

**Comment by avikivity** [comment]
Does the resolver respect the DNS entry TTL?

That's better than yet another special case.

**Comment by dkropachev** [comment]
> Does the resolver respect the DNS entry TTL?
> 
> That's better than yet another special case.

It depends on how exactly you enject it, if you go with unresolved endpoint, then yes, with some problems:
1. It respects it, but it grabs only first AAA record, which means that if first node is down, then you got no connectivity
2. There are two lists, one - initial contact endpoints, and effective cluster nodes, when session is initialized first is merged into second and driver works only with second one, there are two cases when endpoint from `effective cluster nodes` is getting overriden by information from `system.peers` and when it happens driver loses ability to use DNS record completely.

In case of regular endpoints:
1. Driver resolves them 1 to 1 on session initialization, i.e. if you have record that points to whole cluster, driver grabs first AAA record from it and adds resolved address to the initial contact endpoints. As you can imagine TTL is not respected, DNS is not used after initialization .

Overall, in both drivers this part of the driver made very clumsy and very hard to refactor, for example in `3.x` one part of the code uses `hostid` to unique id of the `node` while other part of the driver uses `endpoint` for same purpose leaves very little space to solve these issues in minimalistic and meaningull way.
Any significant change will bring big risk of subtle regression that is going to be hard to catch.
That is why we have reserved to re-resolving under certain confined condition when session is effectively broken.

**Comment by mykaul** [comment]
@dkropachev - I believe this was solved for Java driver 4.x, but not for 3.x? Am I correct? If that's the case, shall we do something about it?

**Comment by dkropachev** [comment]
Solved for both:
scylla-3.x - https://github.com/scylladb/java-driver/pull/344
scylla-4.x - https://github.com/scylladb/java-driver/pull/341

**Comment by mykaul** [comment]
@dkropachev - so can we close this issue?
